test_conformance/subgroups/subhelpers.h

```diff
@@ -22,60 +22,737 @@
 
 #include <limits>
 #include <vector>
-
+#undef min;
+#undef max;
+#define NON_UNIFORM 4
 // Some template helpers
-template <typename Ty> struct TypeName;
-template <> struct TypeName<cl_half> { static const char * val() { return "half"; } };
-template <> struct TypeName<cl_uint> { static const char * val() { return "uint"; } };
-template <> struct TypeName<cl_int> { static const char * val() { return "int"; } };
-template <> struct TypeName<cl_ulong> { static const char * val() { return "ulong"; } };
-template <> struct TypeName<cl_long> { static const char * val() { return "long"; } };
-template <> struct TypeName<float> { static const char * val() { return "float"; } };
-template <> struct TypeName<double> { static const char * val() { return "double"; } };
+namespace subgroups {
+    struct cl_char3 {
+        ::cl_char3 data;
+    };
+    struct cl_uchar3 {
+        ::cl_uchar3 data;
+    };
+    struct cl_short3 {
+        ::cl_short3 data;
+    };
+    struct cl_ushort3 {
+        ::cl_ushort3 data;
+    };
+    struct cl_int3 {
+        ::cl_int3 data;
+    };
+    struct cl_uint3 {
+        ::cl_uint3 data;
+    };
+    struct cl_long3 {
+        ::cl_long3 data;
+    };
+    struct cl_ulong3 {
+        ::cl_ulong3 data;
+    };
+    struct cl_float3 {
+        ::cl_float3 data;
+    };
+    struct cl_double3 {
+        ::cl_double3 data;
+    };
+    struct cl_half {
+        ::cl_half data;
+    };
+    struct cl_half2 {
+        ::cl_half2 data;
+    };
+    struct cl_half3 {
+        ::cl_half3 data;
+    };
+    struct cl_half4 {
+        ::cl_half4 data;
+    };
+    struct cl_half8 {
+        ::cl_half8 data;
+    };
+    struct cl_half16 {
```

Comment:
Why do we need these special types?

Reply:
This is because compilation errors. When one type is used to define another one.
For example the vector cl_uint3 is defined as follows (in cl_platform.h):
typedef  cl_uint4  cl_uint3;
or cl_half is defined as:
typedef unsigned __int16        cl_half;

Using next this types in below templates cause error : type is already defined.